### PR TITLE
MARKET-1636 Video Recorder 

### PR DESCRIPTION
**Pull request by addKiwi**
image

image

### CODE REVIEW

File: components/bl-video-recorder-component/styles/index.less [hunk 1]
```diff
@@ -0,0 +1,13 @@
+@bl-customComponent-videoRecorder-video-background: black;
+@bl-customComponent-videoRecorder-video-width: 720px;
+@bl-customComponent-videoRecorder-video-height: 480px;
```
Comment (serhmel) [inline components/bl-video-recorder-component/styles/index.less:3]: Not sure it's a good idea to assign a fixed size to the component.

File: components/bl-video-recorder-component/README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+<p align="center">
+  <img src="./thumbnail.png" alt="main thumbnail" width="780"/>
+</p>
+
+## Properties
+
+| Property    | Type       | Default Value    | Logic           | Data Binding | UI Setting | Description                                                  |
+|-------------|------------|------------------|-----------------|--------------|------------|--------------------------------------------------------------|
+| Width       | *Number*   | 720              |                 | NO           | YES        | Controls width of component, in 'px'.                        |
+| Height      | *Number*   | 480              |                 | NO           | YES        | Controls height of component, in 'px'.                       |
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/README.md:15]: I assume we should make it bindable for a case if the app could be opened in mobile and desktop

File: components/bl-video-recorder-component/README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+<p align="center">
+  <img src="./thumbnail.png" alt="main thumbnail" width="780"/>
+</p>
+
+## Properties
+
+| Property    | Type       | Default Value    | Logic           | Data Binding | UI Setting | Description                                                  |
+|-------------|------------|------------------|-----------------|--------------|------------|--------------------------------------------------------------|
+| Width       | *Number*   | 720              |                 | NO           | YES        | Controls width of component, in 'px'.                        |
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/README.md:14]: Also, is it possible to use here `%` and other dimensions?

File: components/bl-video-recorder-component/src/helpers.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  }
+};
+
+export const prepareToView = videoRef => {
+  if (videoRef) {
+    videoRef.autoplay = false;
+    videoRef.muted = false;
+    videoRef.controls = true;
+    videoRef.srcObject = null;
+    videoRef.captureStream = null;
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/src/helpers.js:15]: Object.assign?

File: components/bl-video-recorder-component/src/helpers.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  }
+};
+
+export const download = (blob, fileName) => {
+  const link = document.createElement('a');
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/src/helpers.js:20]: empty line after declaration

File: components/bl-video-recorder-component/src/helpers.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  link.href = URL.createObjectURL(blob);
+  link.download = fileName;
+
+  document.body.appendChild(link);
+  link.click();
+
+  setTimeout(() => {
+    URL.revokeObjectURL(link.href);
+    link.parentNode.removeChild(link);
+  }, 0);
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/src/helpers.js:31]: I think it isn't crossplatform? Is there another way to implement download?

File: components/bl-video-recorder-component/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+| Allow Audio | *Checkbox* | `true`           |                 | NO           | YES        | Controls allow recording audio from screen, page, etc.       |
+| File Name   | *Text*     | "Recorded Video" | File Name Logic | YES          | YES        | This is a handler to control the name of the downloaded file.|
+
+## Events
+
+| Name               | Triggers                                  | Context Blocks                            |
+|--------------------|-------------------------------------------|-------------------------------------------|
+| On Start Recording | When the recording is started             |                                           |
+| On Stop Recording  | When the recording is stopped             |                                           |
+| On Download file   | When the recorded file starts downloading | Blob File: `{size: Number, type: String}` |
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/README.md:25]: This blob is not the same as at 34 line?

Also, will be better to add link for MDN than use this signature

File: components/bl-video-recorder-component/src/helpers.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+export const download = (blob, fileName) => {
+  const link = document.createElement('a');
+  link.style.display = 'none';
+  link.href = URL.createObjectURL(blob);
+  link.download = fileName;
+
+  document.body.appendChild(link);
+  link.click();
+
+  setTimeout(() => {
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/src/helpers.js:28]: Why we use timeouts here?

File: components/bl-video-recorder-component/src/helpers.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    URL.revokeObjectURL(link.href);
+    link.parentNode.removeChild(link);
+  }, 0);
+};
+
+export const captureMediaDevices = async (mediaConstraints, videoRef) => {
+  try {
+    const stream = await navigator.mediaDevices.getDisplayMedia(mediaConstraints);
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/src/helpers.js:36]: @Valodya do we need to add this in BackendlessUI?

File: components/bl-video-recorder-component/src/index.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      const chunks = [];
+
+      recorder.current.ondataavailable = event => {
+        if (event.data.size > 0) {
+          chunks.push(event.data);
+        }
+      };
+
+      recorder.current.onstop = () => {
+        const recordedBlob = new Blob(chunks, { type: 'video/webm' });
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/src/index.js:43]: Could we add property for different types, not only webm?

File: components/bl-video-recorder-component/src/index.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+        setRecordedBlob(recordedBlob);
+        onStop();
+        prepareToView(videoRef.current);
+        chunks.length = 0;
+      };
+
+      recorder.current.start(200);
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/src/index.js:51]: Magic number 200. Create constant or get rid of it

File: components/bl-video-recorder-component/src/index.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    }
+  }, [width, height, display]);
+
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <div ref={ elRef } className={ cn('bl-customComponent-videoRecorder', classList) } style={ style }>
+      <video ref={ videoRef } className="video"/>
```
Comment (v-excelsior) [inline components/bl-video-recorder-component/src/index.js:87]: Could we add UI controls that could be hidden/shown?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-video-recorder-component/src/helpers.js (hunk 8 above)
Comment (addKiwi) [reply components/bl-video-recorder-component/src/helpers.js:28]: To make this work on Firefox we need to wait a little while before removing the link.

File: components/bl-video-recorder-component/README.md (hunk 7 above)
Comment (addKiwi) [reply components/bl-video-recorder-component/README.md:25]: The same blob.

File: components/bl-video-recorder-component/src/helpers.js (hunk 6 above)
Comment (addKiwi) [reply components/bl-video-recorder-component/src/helpers.js:31]: I didn't find anything better.

File: components/bl-screen-recorder-component/src/helpers.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+      muted   : true,
+      controls: false,
+    });
+  }
+
+  if (startRef && stopRef) {
+    startRef.disabled = true;
+    stopRef.disabled = false;
+  }
+
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/src/helpers.js:14]: redundant empty line

File: components/bl-screen-recorder-component/src/helpers.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+export const captureMediaDevices = async (mediaConstraints, videoRef) => {
+  try {
+    const stream = await navigator.mediaDevices.getDisplayMedia(mediaConstraints);
+
+    videoRef.current.srcObject = stream;
+
+    return stream;
+  } catch (e) {
+    console.error('The source of the stream did not select.', e);
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/src/helpers.js:59]: `Stream source not selected.` ?

File: components/bl-screen-recorder-component/src/helpers.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+  }
+};
+
+export const ensureMeasure = dimension => {
+  return String(Number(dimension)) === dimension || !isNaN(dimension) ? dimension + 'px' : dimension;
+};
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/src/helpers.js:65]: We haven't something like this in CSSUtils?

File: components/bl-screen-recorder-component/src/index.js [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+  return (
+    <div ref={ elRef } className={ cn('bl-customComponent-screenRecorder', classList) } style={ style }>
+      <video ref={ videoRef } className="video"/>
+      { controls && (
+        <div className="controls">
+          <button ref={ startRef } className="control-button" onClick={ startRecording }>{ startText }</button>
+          <button ref={ stopRef } className="control-button" onClick={ stopRecording }>{ stopText }</button>
+          <button ref={ downloadRef } className="control-button"
+                  onClick={ downloadRecordedFile }>{ downloadText }</button>
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/src/index.js:105]:
```suggestion
<button
  ref={ downloadRef }
  className="control-button"
  onClick={ downloadRecordedFile }>
  { downloadText }
</button>
```

File: components/bl-screen-recorder-component/styles/index.less [hunk 17]
```diff
@@ -0,0 +1,52 @@
+@bl-customComponent-screenRecorder-video-background: black;
+
+@bl-customComponent-screenRecorder-button-backgroundColor: @themePrimary;
+@bl-customComponent-screenRecorder-button-disabled: @disabledColor;
+@bl-customComponent-screenRecorder-button-borderRadius: @appComponentBorderRadius;
+@bl-customComponent-screenRecorder-button-color: @appTextColor;
+@bl-customComponent-screenRecorder-button-shadowColor: @appComponentShadowColor;
+@bl-customComponent-screenRecorder-button-shadowHover: 0px 2px 4px -1px fade(@bl-customComponent-screenRecorder-button-shadowColor, 20%), 0px 4px 5px 0px fade(@bl-customComponent-screenRecorder-button-shadowColor, 14%), 0px 1px 10px 0px fade(@bl-customComponent-screenRecorder-button-shadowColor, 12%);
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/styles/index.less:8]:
```suggestion
@bl-customComponent-screenRecorder-button-shadowHover:
  0px 2px 4px -1px fade(@bl-customComponent-screenRecorder-button-shadowColor, 20%),
  0px 4px 5px 0px fade(@bl-customComponent-screenRecorder-button-shadowColor, 14%),
  0px 1px 10px 0px fade(@bl-customComponent-screenRecorder-button-shadowColor, 12%);
```

Could we do like that?

File: components/bl-screen-recorder-component/styles/index.less [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+  .control-button {
+    cursor: pointer;
+    background: @bl-customComponent-screenRecorder-button-backgroundColor;
+    border-radius: @bl-customComponent-screenRecorder-button-borderRadius;
+    border-width: @bl-customComponent-screenRecorder-button-borderWidth;
+    border-style: @bl-customComponent-screenRecorder-button-borderStyle;
+    box-shadow: @bl-customComponent-screenRecorder-button-shadow;
+    height: 36px;
+    width: 25%;
+    color: #fff;
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/styles/index.less:40]: Variable for color and height

width should not be fixed as content could be long

File: components/bl-screen-recorder-component/styles/index.less [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  .controls {
+    display: flex;
+    justify-content: space-evenly;
+    padding-top: 10px
+  }
+
+  .control-button {
+    cursor: pointer;
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/styles/index.less:32]: variable

File: components/bl-screen-recorder-component/styles/index.less [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    background: @bl-customComponent-screenRecorder-button-backgroundColor;
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/styles/index.less:33]: backgroundColor -> background or change the property

File: components/bl-screen-recorder-component/src/index.js [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <div ref={ elRef } className={ cn('bl-customComponent-screenRecorder', classList) } style={ style }>
+      <video ref={ videoRef } className="video"/>
+      { controls && (
+        <div className="controls">
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/src/index.js:101]: Use please more specific classes 

video -> screenRecorder-video
controls -> screenRecorder-controls

File: components/bl-screen-recorder-component/component.json [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+    {
+      "type": "text",
+      "name": "width",
+      "label": "Width",
+      "showInSettings": true,
+      "defaultValue": "720px",
+      "hasLogicHandler": true,
+      "handlerId": "widthLogic",
+      "handlerLabel": "Width Logic",
+      "handlerDescription": "This handler controls component width.",
```
Comment (v-excelsior) [inline components/bl-screen-recorder-component/component.json:20]: Add please settingTooltip for properties with short description of what they do

**Review by v-excelsior — CHANGES_REQUESTED** [review]

**Pull request merged**